add page: Portal Network to developers/docs/networking-layer (#10289)
ethereum/ethereum-org-website: src/content/developers/docs/networking-layer/portal-network/index.md

---
title: The Portal Network
description: An overview of the Portal Network - an in-development network designed to support low-resource clients.
lang: en
---

> Ethereum runs on nodes. A "node" is any instance of Ethereum client software connected to other computers also running Ethereum software, forming a network. A client is an implementation of Ethereum that verifies data against the protocol rules and keeps the network secure.

Reviewer: We don't want to repeat the basics of what are nodes etc too much here imo. Let's tighten this up and get right to the point of the page.

```suggestion
Ethereum is a network made up of computers that run Ethereum client software. Each of these computers is called a 'node'. The client software allows a node to send and receive data on the Ethereum network, and verifies data against the Ethereum protocol rules. Nodes keep a lot of historical data in their disk storage and add to it when they receive new packets of information, known as blocks, from other nodes on the network. This is necessary for always checking that a node has information consistent with the rest of the network. This means running a node can require a lot of disk space. Some node operations can require a lot of RAM too.

To get around this disk storage problem, 'light' nodes have been developed that request information from full nodes instead of storing it all themselves. However, this means the light node is not independently verifying the information and is trusting another node instead. It also means that full nodes are required to take on extra work to serve those light nodes.

The Portal network is a new networking design for Ethereum that aims to solve the data availability problem for "light" nodes without having to trust or put extra strain on full nodes, by sharing the necessary data in small chunks across the network.

Read more about [nodes and clients]((https://ethereum.org/en/developers/docs/nodes-and-clients/))
```

> Nodes have to run client software, for example, Prysm and Geth, which are vital to the operations of Ethereum.

Reviewer: ```suggestion
```

> Ethereum is a decentralized network that relies on a consensus mechanism called Proof-of-Stake (PoS) to validate and agree upon transactions and the state of the blockchain. Clients are implementations of the Ethereum protocol that enable network participants, known as validators, to contribute their computational resources to secure the network and maintain consensus.

Reviewer: ```suggestion
```

> ## Limitations of the current networking design {#limitations-of-current-network-design}

Reviewer: I've learned from @konopkja that it's great for UX to put the benefits to the user clearly up front. Maybe rename this "Why do we need Portal Network" and try to get straight into why the current networking design is resource-heavy and why that's a problem. Then move on to the new design concept.

```suggestion
## Why do we need the Portal network {#why-do-we-need-portal-network}
```

> Node Operators use clients to store a copy of a part or the entire blockchain and validate transactions and smart contracts. By running their nodes, users can independently verify the state of the network without relying on centralized entities.

Reviewer: ```suggestion
Ethereum nodes store their own full or partial copy of the Ethereum blockchain. This local copy is used to validate transactions and ensure the node is following the correct chain. This locally stored data allows nodes to independently verify that incoming data is valid and correct without needing to trust any other entity.
```

> Clients are currently developed on top of peer-to-peer technology called DevP2P and LibP2P.

Reviewer: Let's expose concepts step-by-step. We can link out to our networking pages later where the reader can find out about devP2P and libP2P

```suggestion
```

> The clients do a great job of running Ethereum; they must work within the design framework of devP2P and libP2P.

Reviewer: ```suggestion
```

![devp2p and libp2p](portal-network-devp2p-libp2p.png)

> This means the current clients need more support in what they can do to help keep network data requirements low.

Reviewer: ```suggestion
This means running nodes can be expensive, because a large amount of disk space has to be dedicated to Ethereum. There are several solutions to this problem on the Ethereum roadmap, including [history expiry](https://ethereum.org/en/roadmap/statelessness/#history-expiry), [state expiry](https://ethereum.org/en/roadmap/statelessness/#state-expiry) and [statelessness](https://ethereum.org/en/roadmap/statelessness/). However, these are likely several years away from being implemented. There are also [light nodes](https://ethereum.org/en/developers/docs/nodes-and-clients/light-clients/) that do not save their own copy of the chain data, they request the data they need from full nodes. However, this means light nodes have to trust full nodes to provide honest data and also stresses the full nodes that have to serve the data the light nodes need.
```

Disk space often poses a significant limitation for node operators, particularly when running full nodes or archive nodes for blockchain networks such as Ethereum. To give you an overview of the storage requirements:

For a full Geth node and a consensus client, it is advisable to have a 2TB hard drive. Geth, the Go implementation of the Ethereum protocol, typically demands more than 650GB of disk space for a snap-synced full node. Snap-sync is a synchronization method that involves downloading snapshots of the Ethereum blockchain instead of processing each block individually.

> A Geth node grows by approximately 14GB per week. To optimize storage utilization, Geth offers a pruning feature that allows the node to discard older blockchain data that is no longer necessary for transaction verification. By pruning, the total storage size can be reduced back to the original 650GB.

Reviewer: deleting as this info was moved up the page
```suggestion
```

> However, it's important to note that pruning limits the node's ability to provide historical data.

Reviewer: deleting as this info was moved up the page
```suggestion
```

> Read more here: https://geth.ethereum.org/docs/getting-started/hardware-requirements

Reviewer: deleting as this info was moved up the page
```suggestion
```

> The design of DevP2P impacts users running a node in many ways:

Reviewer: deleting as this info was moved up the page
```suggestion
```

- High CPU usage

> - High hardrive requirements

Reviewer: ```suggestion
```

> - Bandwidth usage

Reviewer: ```suggestion
```

> - Sync times can be long

Reviewer: ```suggestion
```

> Running Ethereum nodes results in a high entry barrier that affects geographic and economic participation. Trying to build solutions to tackle these issues are not feasible in the current framework.

Reviewer: ```suggestion
```

> ## Why is the Portal Network different? {#why-is-portal-network-different?}

Reviewer: ```suggestion
## How does the Portal Network work? {#how-does-portal-network-work?}
```

Portal decided to rethink the whole design.

“What if we could build outside of these design constraints.”

> Portal aims to build a decentralized peer-to-peer network that enables lightweight access to the Ethereum protocol.

Reviewer: ```suggestion
The idea of the Portal Network is to take the best bits of the current networking stack by enabling JSON-RPC style request/responses to be served through a lightweight DevP2P style peer-to-peer decentralized network.
```

> Portal Network's new design aims to:

Reviewer: ```suggestion
The benefits of this network design are:
```

- Reduce Internet bandwidth usage
- Minimized or zero syncing
- Accessible to resource-constrained devices (<1GB ram, <100mB disk, 1CPU)

> - Residential internet connection

Reviewer: same as point 1?
```suggestion
- Residential internet connection
```

> Portal took the essential five parts of running a client that could be split up. This reduces the burden on the above, allowing users to run a node with whatever computational power they can.

Reviewer: ```suggestion
The diagram below shows the functions of existing clients that can be delivered by the Portal Network, enabling users to access these functions on very low-resource devices.
```

![portal network table](portal-network-table.png)

> Another challenge is building this in a decentralised structure; running on a centralised infrastructure would remove many of the obstacles the team is building solutions for, although this would be too much dependency which would not be a viable solution for the long term.

Reviewer: think this is orthogonal to the point of the section
```suggestion
```

> Helios, a Rust-based Ethereum light client, provides access to Ethereum. converts data from an untrusted centralized provider; while this project overcomes some challenges, it needs to be fully decentralised.

Reviewer: No need to call on Helios here imo - we already make the point that present light clients rely on trusted RPC providers
```suggestion
```

> 

Reviewer: ## Client diversity by default {#client-diversity-as-default}

Portal Network also made the design choice to build three clients from day one.

Having multiple independent client implementations enhances the resilience and decentralization of the Ethereum network.

If one client experiences issues or vulnerabilities, other clients can continue to operate smoothly, preventing a single point of failure. Additionally, diverse client implementations foster innovation and competition, driving improvements and reducing monoculture risk within the ecosystem.

> Now you see that the best option to design a decentralised light client needed a new approach from the current infrastructure, a multi-team effort.

Reviewer: ```suggestion
```

To get around this disk storage problem, 'light' nodes have been developed that request information from full nodes instead of storing it all themselves. However, this means the light node is not independently verifying the information and is trusting another node instead. It also means that full nodes are required to take on extra work to serve those light nodes.

The Portal network is a new networking design for Ethereum that aims to solve the data availability problem for "light" nodes without having to trust or put extra strain on full nodes, by sharing the necessary data in small chunks across the network.

> Read more about [nodes and clients]((https://ethereum.org/en/developers/docs/nodes-and-clients/))

Reviewer 2: ```suggestion
[More on nodes and clients](/developers/docs/nodes-and-clients/)
```

> 
> 

Reviewer 2: ```suggestion
```

## Why do we need the Portal network {#why-do-we-need-portal-network}

Ethereum nodes store their own full or partial copy of the Ethereum blockchain. This local copy is used to validate transactions and ensure the node is following the correct chain. This locally stored data allows nodes to independently verify that incoming data is valid and correct without needing to trust any other entity.

This local copy of the blockchain and associated state and receipt data takes up a lot of space on the node's hard disk. For example, a 2TB hard disk is recommended for running a node  using [Geth](https://geth.ethereum.org) paired to a consensus client. Using snap sync, which only stores chain data from a relatively recent set of blocks,  Geth typically occupies about 650GB of disk space but grows at around 14GB/week (you can prune the node back down to 650GB periodically). 

> 
> 

Reviewer 2: ```suggestion
```

This means running nodes can be expensive, because a large amount of disk space has to be dedicated to Ethereum. There are several solutions to this problem on the Ethereum roadmap, including [history expiry](https://ethereum.org/en/roadmap/statelessness/#history-expiry), [state expiry](https://ethereum.org/en/roadmap/statelessness/#state-expiry) and [statelessness](https://ethereum.org/en/roadmap/statelessness/). However, these are likely several years away from being implemented. There are also [light nodes](https://ethereum.org/en/developers/docs/nodes-and-clients/light-clients/) that do not save their own copy of the chain data, they request the data they need from full nodes. However, this means light nodes have to trust full nodes to provide honest data and also stresses the full nodes that have to serve the data the light nodes need.

Portal Network aims to provide an alternative way for light nodes to get their data that does not require trusting or adding significantly to the work that has to be done by full nodes. The way this will be done is to introduce a new way for Ethereum nodes to share data across the network.

> 

Reviewer 2: ```suggestion
```

> ## How does the Portal Network work? {#how-does-portal-network-work}

Reviewer 2: ```suggestion
## How does the Portal Network work? {#how-does-portal-network-work}

```

The core idea of the Portal Network is to take the best bits of the current networking stack by enabling information needed by light clients, such as historical data and the identity of the current head of the chain to be served through a lightweight DevP2P style peer-to-peer decentralized network using a [DHT](https://en.wikipedia.org/wiki/Distributed_hash_table) (similar to Bittorrent).

The idea is to add small parts of the total historical Ethereum data and some specific node responsibilities to each node. Then, requests are served by seeking out the nodes storing the specific data that was requested and retrieving it from them. 

This inverts the normal model of light nodes finding a single node and requesting them to filter and serve large volumes of data; instead, they quickly filter a large network of nodes that each handle small amounts of data.

The goal is to allow a decentralized network of lightweight Portal clients to:

- track the head of the chain
- sync recent and historical chain data
- retrieve state data
- broadcast transactions
- execute transactions using the [EVM](https://ethereum.org/en/developers/docs/evm/)

> The benefits of this network design are:

Reviewer 2: ```suggestion
The benefits of this network design are:

```

- reduce dependence on centralized providers
- Reduce Internet bandwidth usage
- Minimized or zero syncing
- Accessible to resource-constrained devices (<1GB ram, <100mB disk, 1CPU)

The diagram below shows the functions of existing clients that can be delivered by the Portal Network, enabling users to access these functions on very low-resource devices.

![portal network table](portal-network-table2.png)

## Client diversity by default {#client-diversity-as-default}

The Portal Network developers also made the design choice to build three separate Portal Network clients from day one.

> The clients are:

Reviewer 2: ```suggestion
The clients are:

```

- [Trin](https://github.com/ethereum/trin): written in Rust
- [Fluffy](https://nimbus.team/docs/fluffy.html): written in Nim
- [Ultralight](https://github.com/ethereumjs/ultralight): written in Typescript

Having multiple independent client implementations enhances the resilience and decentralization of the Ethereum network.

If one client experiences issues or vulnerabilities, other clients can continue to operate smoothly, preventing a single point of failure. Additionally, diverse client implementations foster innovation and competition, driving improvements and reducing monoculture risk within the ecosystem.

> ## Further Reading {#futher-reading}

Reviewer 2: ```suggestion
## Further reading {#futher-reading}
```

![devP2P and libP2P](portal-network-devp2p-libp2p.png)

Nodes can also serve specific data through the [JSON-RPC API](https://ethereum.org/en/developers/docs/apis/json-rpc/), which is how apps and wallets swap information with Ethereum nodes. However, none of these are ideal protocols for serving data to light clients. 

Light clients can't currently request specific pieces of chain data over DevP2P or libP2p because those protocols are only designed to enable chain synchronization and gossiping of blocks and transactions. Light clients do not want to download this information because that would stop them from being "light".

The JSON-RPC API is not an ideal choice for light client data requests either, because it relies upon a connection to a specific full node or centralized RPC provider that can serve the data. This means the light client has to trust that specific node/provider to be honest, and also the full node might have to handle lots of requests from many light clients, adding to their bandwidth requirements.

> Portal decided to rethink the whole design.

Reviewer: ```suggestion
The point of the Portal Network is to rethink the whole design, building specifically for lightness, outside of the design constraints of the existing Ethereum clients.
```

> “What if we could build outside of these design constraints.”

Reviewer: ```suggestion
```

- Minimized or zero syncing
- Accessible to resource-constrained devices (<1GB ram, <100mB disk, 1CPU)

The diagram below shows the functions of existing clients that can be delivered by the Portal Network, enabling users to access these functions on very low-resource devices.

![portal network table](portal-network-table2.png)

## Client diversity by default {#client-diversity-as-default}

The Portal Network developers also made the design choice to build three separate Portal Network clients from day one.
The clients are:
- [Trin](https://github.com/ethereum/trin): written in Rust
- [Fluffy](https://nimbus.team/docs/fluffy.html): written in Nim
- [Ultralight](https://github.com/ethereumjs/ultralight): written in Typescript

Having multiple independent client implementations enhances the resilience and decentralization of the Ethereum network.

If one client experiences issues or vulnerabilities, other clients can continue to operate smoothly, preventing a single point of failure. Additionally, diverse client implementations foster innovation and competition, driving improvements and reducing monoculture risk within the ecosystem.

## Further Reading {#futher-reading}

> - The Portal Network (Piper Merriam at ETHZurich).

Reviewer: ```suggestion
- [The Portal Network (Piper Merriam at Devcon Bogota)](https://www.youtube.com/watch?v=0stc9jnQLXA).
```

Read more about [nodes and clients]((https://ethereum.org/en/developers/docs/nodes-and-clients/))

## Why do we need the Portal network {#why-do-we-need-portal-network}

Ethereum nodes store their own full or partial copy of the Ethereum blockchain. This local copy is used to validate transactions and ensure the node is following the correct chain. This locally stored data allows nodes to independently verify that incoming data is valid and correct without needing to trust any other entity.

This local copy of the blockchain and associated state and receipt data takes up a lot of space on the node's hard disk. For example, a 2TB hard disk is recommended for running a node  using [Geth](https://geth.ethereum.org) paired to a consensus client. Using snap sync, which only stores chain data from a relatively recent set of blocks,  Geth typically occupies about 650GB of disk space but grows at around 14GB/week (you can prune the node back down to 650GB periodically). 

> ![devp2p and libp2p](portal-network-devp2p-libp2p.png)

Reviewer: rm duplicate
```suggestion
```

This means running nodes can be expensive, because a large amount of disk space has to be dedicated to Ethereum. There are several solutions to this problem on the Ethereum roadmap, including [history expiry](https://ethereum.org/en/roadmap/statelessness/#history-expiry), [state expiry](https://ethereum.org/en/roadmap/statelessness/#state-expiry) and [statelessness](https://ethereum.org/en/roadmap/statelessness/). However, these are likely several years away from being implemented. There are also [light nodes](https://ethereum.org/en/developers/docs/nodes-and-clients/light-clients/) that do not save their own copy of the chain data, they request the data they need from full nodes. However, this means light nodes have to trust full nodes to provide honest data and also stresses the full nodes that have to serve the data the light nodes need.

Portal Network aims to provide an alternative way for light nodes to get their data that does not require trusting or adding significantly to the work that has to be done by full nodes. The way this will be done is to introduce a new way for Ethereum nodes to share data across the network.

> ## How does the Portal Network work? {#how-does-portal-network-work?}

Reviewer: think this will fix the header tag
```suggestion
## How does the Portal Network work? {#how-does-portal-network-work}
```

Portal Network aims to provide an alternative way for light nodes to get their data that does not require trusting or adding significantly to the work that has to be done by full nodes. The way this will be done is to introduce a new way for Ethereum nodes to share data across the network.

## How does the Portal Network work? {#how-does-portal-network-work?}
Ethereum nodes have strict protocols that define how they communicate with each other. Execution clients communicate using a set of subprotocols known as [DevP2P](https://ethereum.org/en/developers/docs/networking-layer/#devp2p), while consensus clients use a different stack of subprotocols called [libP2P](https://ethereum.org/en/developers/docs/networking-layer/#libp2p). These define the types of data that can be passed between nodes. 

![devP2P and libP2P](portal-network-devp2p-libp2p.png)

Nodes can also serve specific data through the [JSON-RPC API](https://ethereum.org/en/developers/docs/apis/json-rpc/), which is how apps and wallets swap information with Ethereum nodes. However, none of these are ideal protocols for serving data to light clients. 

> Light client's can't currently request specific pieces of chain data over DevP2P or libP2p because those protocols are only designed to enable chain synchronization and gossiping of blocks and transactions. Light clients do not want to download this information because that would stop them from being "light".

Reviewer 3: `Light clients can't currently ...`

Reviewer: ```suggestion
Light clients can't currently request specific pieces of chain data over DevP2P or libP2p because those protocols are only designed to enable chain synchronization and gossiping of blocks and transactions. Light clients do not want to download this information because that would stop them from being "light".
```

The JSON-RPC API is not an ideal choice for light client data requests either, because it relies upon a connection to a specific full node or centralized RPC provider that can serve the data. This means the light client has to trust that specific node/provider to be honest, and also the full node might have to handle lots of requests from many light clients, adding to their bandwidth requirements.

Portal decided to rethink the whole design.

“What if we could build outside of these design constraints.”

> The core idea of the Portal Network is to take the best bits of the current networking stack by enabling information needed by light clients, such as historical data and the identity of the current head of the chain to be served through a lightweight DevP2P style peer-to-peer decentralized network.

Reviewer 3: "... to be served through a lightweight, DevP2p style, peer-to-peer, decentralized network using a [DHT](https://en.wikipedia.org/wiki/Distributed_hash_table) (similar to Bittorrent)."

Reviewer: ```suggestion
The core idea of the Portal Network is to take the best bits of the current networking stack by enabling information needed by light clients, such as historical data and the identity of the current head of the chain to be served through a lightweight DevP2P style peer-to-peer decentralized network using a [DHT](https://en.wikipedia.org/wiki/Distributed_hash_table) (similar to Bittorrent).
```

The idea is to add small parts of the total historical Ethereum data and some specific node responsibilities to each node. Then, requests are served by seeking out the nodes storing the specific data that was requested and retrieving it from them. 

This inverts the normal model of light nodes finding a single node and requesting them to filter and serve large volumes of data; instead, they quickly filter a large network of nodes that each handle small amounts of data.

The goal is to allow a decentralized network of lightweight Portal clients to:

- track the head of the chain
- sync recent and historical chain data
- retrieve state data
- broadcast transactions
- execute transactions using the [EVM](https://ethereum.org/en/developers/docs/evm/)

The benefits of this network design are:

> 

Reviewer 3: It's sort of implicit in the "decentralization" feature, but it could be worth listing something about "minimizing dependence on centralized providers" here.

Reviewer: ```suggestion
- reduce dependence on centralized providers
```

- Reduce Internet bandwidth usage
- Minimized or zero syncing
- Accessible to resource-constrained devices (<1GB ram, <100mB disk, 1CPU)

> - Residential internet connection

Reviewer: ```suggestion
```

The diagram below shows the functions of existing clients that can be delivered by the Portal Network, enabling users to access these functions on very low-resource devices.

> ![portal network table](portal-network-table.png)

Reviewer 3: AFAIK the plan is currently to deprecate building out the `Canonical Txn Index`

## Client diversity by default {#client-diversity-as-default}

The Portal Network developers also made the design choice to build three separate Portal Network clients from day one.

> 

Reviewer: ```suggestion
The clients are:
- [Trin](https://github.com/ethereum/trin): written in Rust
- [Fluffy](https://nimbus.team/docs/fluffy.html): written in Nim
- [Ultralight](https://github.com/ethereumjs/ultralight): written in Typescript

```

Having multiple independent client implementations enhances the resilience and decentralization of the Ethereum network.

> If one client experiences issues or vulnerabilities, other clients can continue to operate smoothly, preventing a single point of failure. Additionally, diverse client implementations foster innovation and competition, driving improvements and reducing monoculture risk within the ecosystem.

Reviewer 3: Is it worth linking to the github repo for each client here?

Reviewer: yes, sure, coming up...
